libraries/tools/idl2k/src/main/kotlin/build.kt

```diff
@@ -0,0 +1,129 @@
+package org.jetbrains.idl2k
+
+import org.antlr.v4.runtime.ANTLRFileStream
+import kotlin.collections.HashSet
+import java.io.File
+import java.io.IOException
+import java.net.URL
+import java.util.ArrayList
+import java.util.LinkedHashMap
+
+
+class BuildWebIdl(val mdnCacheFile: File, val srcDir: File) {
+    val repositoryPre = loadPreliminaryRepository()
+
+    fun loadPreliminaryRepository(): Repository {
+        if (!srcDir.exists()) {
+            System.err?.println("Directory ${srcDir.absolutePath} doesn't exist")
+            System.exit(1)
+        }
+
+        return srcDir.walkTopDown().filter { it.isDirectory || it.extension == "idl" }.asSequence().filter { it.isFile }.toList().sortedBy { it.absolutePath }.fold(Repository(emptyMap(), emptyMap(), emptyMap(), emptyMap())) { acc, e ->
+            System.err.flush()
+            System.err.println("Parsing ${e.absolutePath}")
+            val fileRepository = parseIDL(ANTLRFileStream(e.absolutePath, "UTF-8"))
+
+            Repository(
+                    interfaces = acc.interfaces.mergeReduce(fileRepository.interfaces, ::merge),
+                    typeDefs = acc.typeDefs + fileRepository.typeDefs,
+                    externals = acc.externals.merge(fileRepository.externals),
+                    enums = acc.enums + fileRepository.enums
+            )
+        }
+    }
+
+    init {
+        println("Prepare...")
+    }
+
+    val repository = repositoryPre.copy(typeDefs = repositoryPre.typeDefs.mapValues { it.value.copy(mapType(repositoryPre, it.value.types)) })
+
+    val definitions = mapDefinitions(repository, repository.interfaces.values).map {
+        if (it.name in relocations) {
+            // we need this to get interfaces listed in the relocations in valid package
+            // to keep compatibility with DOM Java API
+            it.copy(namespace = relocations[it.name]!!)
+        } else {
+            it
+        }
+    }
+    val unions = generateUnions(definitions, repository.typeDefs.values)
+
+    init {
+        implementInterfaces(definitions)
+    }
+
+    val allPackages = (definitions.asSequence().map { it.namespace } + repository.enums.values.map { it.namespace }).distinct().sorted()
+
+    val mdnCache = updateMdnCache()
```

Comment:
Populating MDN cache over internet in the constructor... well,  I'd extract it to an additional function `prepare()`

Reply:
I made `mdnCache` `by lazy`.